docs/proposals/003-atomic-t.md

```diff
@@ -0,0 +1,104 @@
+SP #003 - `Atomic<T>` type
+==============
+
+
+Status
+------
+
+Author: Yong He
+
+Status: Design Discussion.
+
+Implementation: N/A
+
+Reviewed by: N/A
+
+Background
+----------
+
+HLSL defines atomic intrinsics to work on free references to ordinary values such as `int` and `float`. However, this doesn't translate well to Metal and WebGPU,
+which defines `atomic<T>` type and only allow atomic operations to be applied on values of `atomic<T>` types.
+
+Slang's Metal backend follows the same technique in SPIRV-Cross and DXIL->Metal converter that relies on a C++ undefined behavior that casts an ordinary `int*` pointer to a `atomic<int>*` pointer
+and then call atomic intrinsic on the reinterpreted pointer. This is fragile and not guaranteed to work in the future.
+
+To make the situation worse, WebGPU bans all possible ways to cast a normal pointer into an `atomic` pointer. In order to provide a truly portable way to define
+atomic operations and allow them to be translatable to all targets, we will also need an `atomic<T>` type in Slang that maps to `atomic<T>` in WGSL and Metal, and maps to
+`T` for HLSL/SPIRV.
+
+
+Proposed Approach
+-----------------
+
+We define an `Atomic<T>` type that functions as a wrapper of `T` and provides atomic operations:
+```
+interface IAtomicable {}
+extension int : IAtomicable {}
+extension uint : IAtomicable {}
+extension float : IAtomicable {}
+extension half : IAtomicable {}
+
+struct Atomic<T>
+{
+    T load();
+    [ref] void store(T newValue); // Question: do we really need this?
+    [ref] T exchange(T newValue); // returns old value
+    [ref] T compareExchange(T compareValue, T newValue); // returns old value.
+    [ref] T atomicAdd(T value); // returns original value
+    [ref] T atomicSub(T value); // returns original value
+    [ref] T atomicMax(T value); // returns original value
+    [ref] T atomicMin(T value); // returns original value
+}
+
+extension<T> Atomic<T>
+    where T : IAtomicable
+    where T : __BuiltinIntegerType
```

Comment:
I get what this is doing, but I also worry a little bit, when I think about how this would translate to my "move the required operations into the interface" approach. Effectively, this is saying that any type that conforms `IAtomicable & __BuiltinIntegerType` has *additional* requirements beyond those that are stated for either interface alone.

I really do think the right answer here will be a hierarchy of interfaces, along the lines of:

```
interface IAtomicable { ... } // load/store and exchange
interface IAtomicCompareExchangeable : IAtomicable { ... } // compare-exchange
interface IAtomicNumeric : IAtomicCompareExchangeable { ... } // add/sub/min/max
interface IAtomicLogical : IAtomicCompareExchangeable { ... } // and/or/xor
typealias IAtomicInteger = IAtomicNumeric & IAtomicLogical;
```

(TBD whether add/sub and min/max should be separated from one another)

(Note: I did not make `IAtomicLogical` inherit from `IAtomicNumeric` because it is in principle possible to support `Atomic<bool>`, which would have logical operations but not add/sub/min/max)

It is likely to be rare for programmers to define their own generics that work with `Atomic<T>` for various `T`, so putting the burden on them to specify the range of atomic operations they need to be able to perform doesn't seem like too much.

That said, it is clear that my little hierarchy above closely mirrors the kind of hierarchy we need for builtin scalar types, so it is potentially frustrating to have both `IWhatever` and `IAtomicWhatever` as distinct interfaces.

Reply:
Used hierarchy of IAtomic to clean this up.